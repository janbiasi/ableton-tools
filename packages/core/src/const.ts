import { homedir } from 'os';
import { resolve } from 'path';

export const ENCODING = 'utf8';

export const HOME_DIR = homedir();

export const ABLETON_TOOLS_ROOT = resolve(HOME_DIR, '.abletontools');

export const VERSION = require('../package.json').version;
