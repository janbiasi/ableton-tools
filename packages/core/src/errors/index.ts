import { BackupRestorationError } from './BackupRestorationError';
import { InvariantViolationError } from './InvariantViolationError';
import { ProjectNotFoundError } from './ProjectNotFoundError';
import { ProjectMetaNotFoundError } from './ProjectMetaNotFoundError';
import { DatabaseError } from './DatabaseError';

export {
	BackupRestorationError,
	InvariantViolationError,
	ProjectNotFoundError,
	ProjectMetaNotFoundError,
	DatabaseError,
};
