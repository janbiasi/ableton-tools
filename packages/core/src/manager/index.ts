export { Manager } from './Manager';
export { TemplateManager } from './TemplateManager';
export { ProjectManager } from './ProjectManager';
