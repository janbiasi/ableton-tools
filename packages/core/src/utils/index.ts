import { findLiveSetsInFolder } from './als';
import { createArchive, unpackArchive } from './compression';
import { fileExists, folderExists, getLastPathFragment } from './fs';
import { invariant } from './invariant';
import { safeDirName, formatBytes } from './string';
import { uuid } from './uuid';

export { findLiveSetsInFolder, createArchive, unpackArchive, fileExists, folderExists, getLastPathFragment, invariant, safeDirName, formatBytes, uuid };
