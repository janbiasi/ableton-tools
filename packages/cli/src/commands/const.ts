import { Runtime } from "@ableton-tools/core";

export type BuildCommand = (runtime: Runtime) => (...args: any[]) => Promise<any>;
