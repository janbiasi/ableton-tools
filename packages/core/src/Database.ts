import * as fs from 'fs-extra';
import { resolve } from 'path';
import { DatabaseError } from './errors';
import { uuid, fileExists } from './utils';
import { ENCODING, ABLETON_TOOLS_ROOT } from './const';

export interface IDatabaseConfig {
	name: string;
	root?: string;
	fileName?: string;
}

export interface IDatabaseEntry<T> {
	id: string;
	timestamp: number;
	data: T;
}

export interface IDatabaseContent<T extends { [key: string]: any }> {
	name: string;
	entries: IDatabaseEntry<T>[];
}

export class Database<T extends { [key: string]: any }> {
	private contents: IDatabaseContent<T> = {
		name: 'unknown',
		entries: [],
	};

	constructor(private config: IDatabaseConfig) {}

	public async init() {
		this.contents = {
			name: this.config.name,
			entries: [],
		};

		// ensure base directory exists
		await fs.mkdirp(this.root);

		// get initial database contents and safe DB initially
		await this.syncFromDatabase();
		await this.syncToDatabase();
	}

	public get entries() {
		return this.contents.entries as Readonly<IDatabaseEntry<T>[]>;
	}

	public async reset() {
		this.contents = {
			name: this.contents.name,
			entries: [],
		};

		await this.syncToDatabase();
	}

	/**
	 * Update or create an entry
	 */
	public async save(entry: T) {
		const index = this.contents.entries.push({
			id: uuid(),
			timestamp: Date.now(),
			data: entry,
		});

		await this.syncToDatabase();

		return this.contents.entries[index - 1];
	}

	public async deleteByID(id: string) {
		const entry = this.getById(id);

		if (entry) {
			let deletedItem: IDatabaseEntry<T> = {} as IDatabaseEntry<T>;

			this.contents.entries = this.entries.reduce((prev, entry) => {
				const itemOrNothing = entry.id === id ? [] : [entry];
				entry.id === id ? (deletedItem = entry) : void 0;

				return [...prev, ...itemOrNothing];
			}, [] as IDatabaseEntry<T>[]);

			await this.syncToDatabase();

			return deletedItem;
		}

		return false;
	}

	/**
	 * Generic getter for entries using a custom lookup function
	 */
	public get(lookup: (entry: IDatabaseEntry<T>) => boolean): IDatabaseEntry<T>[] {
		return this.contents.entries.filter(lookup);
	}

	/**
	 * Generic getter for one entry using a custom lookup function
	 */
	public getOne(lookup: (entry: IDatabaseEntry<T>) => boolean): IDatabaseEntry<T> | undefined {
		const [entry] = this.get(lookup);
		return entry;
	}

	/**
	 * Get an entry by ID
	 */
	public getById(id: string): IDatabaseEntry<T> | undefined {
		return this.contents.entries.find((entry): boolean => entry.id === id);
	}

	public async syncToDatabase() {
		try {
			await fs.writeFile(this.databaseFile, this.serialize());
		} catch (err) {
			throw new DatabaseError(this.config.name, `Sync to database has failed: ${err.message}`);
		}
	}

	public async syncFromDatabase() {
		const doesDatabaseExist = await this.checkDatabaseExists();

		if (!doesDatabaseExist) {
			// no database found, return current state
			return;
		}

		try {
			// database found, try reading and persisting it
			const dehydrated = await fs.readFile(this.databaseFile, ENCODING);
			this.contents = this.deserialize(dehydrated);
		} catch (err) {
			throw new DatabaseError(this.config.name, `Could not read database: ${err.message}`);
		}
	}

	public toString() {
		return `Database(${this.config.name}) { ${this.contents.entries.length} }`;
	}

	private get root() {
		return this.config.root || ABLETON_TOOLS_ROOT;
	}

	private get databaseFile() {
		return resolve(this.root, `${this.config.fileName || this.config.name}.json`);
	}

	private async checkDatabaseExists() {
		return await fileExists(this.databaseFile);
	}

	private serialize() {
		return JSON.stringify(this.contents, null, 2);
	}

	private deserialize(contents: string): IDatabaseContent<any> {
		return JSON.parse(contents);
	}
}
