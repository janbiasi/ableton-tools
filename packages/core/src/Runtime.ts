import { resolve } from 'path';
import { ABLETON_TOOLS_ROOT, VERSION } from './const';
import { TemplateManager } from './manager/TemplateManager';
import { ProjectManager } from './manager';

export interface IRuntimeConfig {
	root: string;
	version: string;
}

export const defaultConfig: IRuntimeConfig = {
	root: ABLETON_TOOLS_ROOT,
	version: VERSION,
};

export class Runtime {
	public config: IRuntimeConfig;
	public templates: TemplateManager;
	public projects: ProjectManager;

	constructor(config: Partial<IRuntimeConfig> = {}) {
		this.config = { ...defaultConfig, ...config };
		this.templates = new TemplateManager(this.config.root);
		this.projects = new ProjectManager(this.config.root);
	}

	public async bootstrap() {
		await this.templates.init();
		await this.projects.init();

		return this;
	}
}
