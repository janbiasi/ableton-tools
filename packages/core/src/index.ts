import { ENCODING, ABLETON_TOOLS_ROOT } from './const';
import { IProjectSchema, IProjectBackupSchema, IProjectMetaSchema } from './schema/Project';
import { ITemplateSchema } from './schema/Template';
import { ProjectManager, PROJECT_MANAGER_META_FILE } from './manager/ProjectManager';
import { TemplateManager } from './manager';
import { Runtime } from './Runtime';
import { Database } from './Database';
import {
	BackupRestorationError,
	InvariantViolationError,
	ProjectNotFoundError,
	ProjectMetaNotFoundError,
	DatabaseError,
} from './errors';
import {
	findLiveSetsInFolder,
	createArchive,
	unpackArchive,
	fileExists,
	folderExists,
	getLastPathFragment,
	invariant,
	safeDirName,
	formatBytes,
	uuid,
} from './utils';

export {
	// typedefs
	ITemplateSchema,
	IProjectSchema,
	IProjectBackupSchema,
	IProjectMetaSchema,
	// constants
	ENCODING,
	ABLETON_TOOLS_ROOT,
	PROJECT_MANAGER_META_FILE,
	// core
	Runtime,
	Database,
	ProjectManager,
	TemplateManager,
	// utilities
	findLiveSetsInFolder,
	createArchive,
	unpackArchive,
	fileExists,
	folderExists,
	getLastPathFragment,
	invariant,
	safeDirName,
	formatBytes,
	uuid,
	// errors
	BackupRestorationError,
	InvariantViolationError,
	ProjectNotFoundError,
	ProjectMetaNotFoundError,
	DatabaseError,
};
