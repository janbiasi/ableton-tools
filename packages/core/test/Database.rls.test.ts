import { resolve } from 'path';
import { readFile } from 'fs-extra';
import { Database, IDatabaseContent } from '../src/Database';
import { ENCODING } from '../src/const';
import { TEMP_TARGET_DIR } from './const';

const ROOT_DIR = resolve(TEMP_TARGET_DIR, 'database');

const getTestDbContents = async <T extends any>(name: string): Promise<IDatabaseContent<T> & { size: number }> => {
	const finalPath = resolve(ROOT_DIR, `${name}.json`);
	const rawContent = await readFile(finalPath, ENCODING);
	const content: IDatabaseContent<any> = JSON.parse(rawContent);

	return {
		name: content.name,
		entries: content.entries.map((entry) => {
			return {
				...entry,
				id: 'entry-id',
				timestamp: 0,
			};
		}),
		size: content.entries.length,
	};
};

describe('Core > Database (RLS)', () => {
	it('should work correctly', async () => {
		const db = new Database<{ value: string }>({
			root: ROOT_DIR,
			fileName: 'test-db',
			name: 'RLS Test Case',
		});

		await db.init();
		await db.reset();

		const contentAfterInit = await getTestDbContents('test-db');
		expect(contentAfterInit.entries).toHaveLength(0);
		expect(contentAfterInit).toMatchSnapshot();

		const savedItem = await db.save({
			value: 'Hello world!',
		});
		const contentAfterSave = await getTestDbContents('test-db');
		expect(contentAfterSave.entries).toHaveLength(1);
		expect(contentAfterSave).toMatchSnapshot();

		await db.deleteByID(savedItem.id);
		const contentAfterDelete = await getTestDbContents('test-db');
		expect(contentAfterDelete.entries).toHaveLength(0);
		expect(contentAfterDelete).toMatchSnapshot();
	});
});
