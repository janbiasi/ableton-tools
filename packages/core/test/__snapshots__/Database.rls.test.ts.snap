// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Core > Database (RLS) should work correctly 1`] = `
Object {
  "entries": Array [],
  "name": "RLS Test Case",
  "size": 0,
}
`;

exports[`Core > Database (RLS) should work correctly 2`] = `
Object {
  "entries": Array [
    Object {
      "data": Object {
        "value": "Hello world!",
      },
      "id": "entry-id",
      "timestamp": 0,
    },
  ],
  "name": "RLS Test Case",
  "size": 1,
}
`;

exports[`Core > Database (RLS) should work correctly 3`] = `
Object {
  "entries": Array [],
  "name": "RLS Test Case",
  "size": 0,
}
`;
